File: packages/domain/src/extraction/parser.ts
/**
 * Deterministic line parser and manual-entry constructor for atomic claim drafts.
 */
import type {
  AtomicityAssessment,
  ClaimDraft,
  ExtractionUncertainty,
  ManualClaimEntry,
  ParsedClaimLine,
} from './types.js';

const CLAUSE_SEPARATOR = /(?:[.;!?]\s+|\s+(?:and|or|but)\s+)/iu;

function nonEmpty(value: string, label: string): string {
  const trimmed = value.trim();
  if (!trimmed) throw new Error(`${label} is required`);
  return trimmed;
}

function splitEscapedPipe(line: string): string[] {
  const fields: string[] = [];
  let current = '';
  let escaped = false;
  for (const character of line) {
    if (escaped) {
      current += character;
      escaped = false;
    } else if (character === '\\') {
      escaped = true;
    } else if (character === '|') {
      fields.push(current.trim());
      current = '';
    } else {
      current += character;
    }
  }
  if (escaped) current += '\\';
  fields.push(current.trim());
  return fields;
}

export function assessAtomicity(predicate: string, object: string): AtomicityAssessment {
  const combined = `${predicate.trim()} ${object.trim()}`;
  const separators = combined.match(new RegExp(CLAUSE_SEPARATOR.source, 'giu')) ?? [];
  const assertionCount = separators.length + 1;
  return {
    assertionCount,
    independentlySupportable: assertionCount === 1,
    rationale:
      assertionCount === 1
        ? 'No deterministic multi-assertion separator was detected.'
        : `Detected ${String(separators.length)} possible assertion separator(s).`,
  };
}

/**
 * Parse one claim per non-comment line using:
 * `entityId | predicate | object`. A literal pipe is escaped as `\|`.
 */
export function parseClaimLines(input: string): ParsedClaimLine[] {
  const parsed: ParsedClaimLine[] = [];
  for (const [index, rawLine] of input.replace(/\r\n?/gu, '\n').split('\n').entries()) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#')) continue;
    const fields = splitEscapedPipe(line);
    if (fields.length !== 3) {
      throw new Error(`Line ${String(index + 1)} must contain exactly three pipe-delimited fields`);
    }
    const [rawEntityId, rawPredicate, rawObject] = fields as [string, string, string];
    const entityId = nonEmpty(rawEntityId, `Line ${String(index + 1)} entityId`);
    const predicate = nonEmpty(rawPredicate, `Line ${String(index + 1)} predicate`);
    const object = nonEmpty(rawObject, `Line ${String(index + 1)} object`);
    const atomicity = assessAtomicity(predicate, object);
    const uncertainties: ExtractionUncertainty[] = atomicity.independentlySupportable
      ? []
      : [
          {
            code: 'atomicity',
            detail: atomicity.rationale,
            recordedBy: 'parser',
          },
        ];
    parsed.push({
      lineNumber: index + 1,
      draft: { entityId, predicate, object },
      uncertainties,
    });
  }
  return parsed;
}

export function createManualClaimEntry(input: {
  readonly id: string;
  readonly extractedAt: string;
  readonly extractedBy: string;
  readonly draft: ClaimDraft;
  readonly uncertainties: readonly ExtractionUncertainty[];
}): ManualClaimEntry {
  return {
    id: nonEmpty(input.id, 'Manual entry id'),
    extractedAt: nonEmpty(input.extractedAt, 'Manual extraction timestamp'),
    extractedBy: nonEmpty(input.extractedBy, 'Manual extractor'),
    draft: {
      ...input.draft,
      claimId: nonEmpty(input.draft.claimId, 'Claim id'),
      claimVersionId: nonEmpty(input.draft.claimVersionId, 'Claim version id'),
      entityId: nonEmpty(input.draft.entityId, 'Entity id'),
      predicate: nonEmpty(input.draft.predicate, 'Predicate'),
      object: nonEmpty(input.draft.object, 'Object'),
    },
    uncertainties: [...input.uncertainties],
  };
}

File: workers/research/src/black_book_research/extraction/parser.py
"""Deterministic pipe-delimited atomic-claim parser."""

from __future__ import annotations

import re

from .models import AtomicityAssessment, ExtractionUncertainty, ParsedClaimLine

_CLAUSE_SEPARATOR = re.compile(r"(?:[.;!?]\s+|\s+(?:and|or|but)\s+)", re.IGNORECASE)


def assess_atomicity(predicate: str, object_value: str) -> AtomicityAssessment:
    separators = _CLAUSE_SEPARATOR.findall(f"{predicate.strip()} {object_value.strip()}")
    assertion_count = len(separators) + 1
    return AtomicityAssessment(
        assertion_count=assertion_count,
        independently_supportable=assertion_count == 1,
        rationale=(
            "No deterministic multi-assertion separator was detected."
            if assertion_count == 1
            else f"Detected {len(separators)} possible assertion separator(s)."
        ),
    )


def _split_escaped_pipe(line: str) -> list[str]:
    fields: list[str] = []
    current: list[str] = []
    escaped = False
    for character in line:
        if escaped:
            current.append(character)
            escaped = False
        elif character == "\\":
            escaped = True
        elif character == "|":
            fields.append("".join(current).strip())
            current = []
        else:
            current.append(character)
    if escaped:
        current.append("\\")
    fields.append("".join(current).strip())
    return fields


def parse_claim_lines(value: str) -> tuple[ParsedClaimLine, ...]:
    parsed: list[ParsedClaimLine] = []
    for line_number, raw_line in enumerate(
        value.replace("\r\n", "\n").replace("\r", "\n").split("\n"), start=1
    ):
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        fields = _split_escaped_pipe(line)
        if len(fields) != 3:
            raise ValueError(
                f"Line {line_number} must contain exactly three pipe-delimited fields"
            )
        entity_id, predicate, object_value = fields
        if not entity_id or not predicate or not object_value:
            raise ValueError(f"Line {line_number} claim fields must be non-empty")
        atomicity = assess_atomicity(predicate, object_value)
        uncertainties = (
            ()
            if atomicity.independently_supportable
            else (
                ExtractionUncertainty(
                    code="atomicity",
                    detail=atomicity.rationale,
                    recorded_by="parser",
                ),
            )
        )
        parsed.append(
            ParsedClaimLine(
                line_number=line_number,
                entity_id=entity_id,
                predicate=predicate,
                object=object_value,
                uncertainties=uncertainties,
            )
        )
    return tuple(parsed)
